src/core/stream.c

```diff
@@ -713,7 +706,81 @@ QuicStreamParamGet(
+        QUIC_CONNECTION* Connection = Stream->Connection;
+        Timing->Reason = QUIC_FLOW_BLOCKED_STREAM_ID_FLOW_CONTROL;
+        Timing->Time = Stream->BlockedTimings.StreamIdFlowControl.CumulativeTime;
+        if (Stream->BlockedTimings.StreamIdFlowControl.LastStartTime != 0) {
+            Timing->Time +=
+                CxPlatTimeDiff64(Stream->BlockedTimings.StreamIdFlowControl.LastStartTime, Now);
+        }
+        ++Timing;
+        Timing->Reason = QUIC_FLOW_BLOCKED_STREAM_FLOW_CONTROL;
+        Timing->Time = Stream->BlockedTimings.FlowControl.CumulativeTime;
+        if (Stream->BlockedTimings.FlowControl.LastStartTime != 0) {
+            Timing->Time +=
+                CxPlatTimeDiff64(Stream->BlockedTimings.FlowControl.LastStartTime, Now);
+        }
+        ++Timing;
+        Timing->Reason = QUIC_FLOW_BLOCKED_APP;
+        Timing->Time = Stream->BlockedTimings.App.CumulativeTime;
+        if (Stream->BlockedTimings.App.LastStartTime != 0) {
+            Timing->Time +=
+                CxPlatTimeDiff64(Stream->BlockedTimings.App.LastStartTime, Now);
+        }
+        ++Timing;
+        Timing->Reason = QUIC_FLOW_BLOCKED_SCHEDULING;
+        Timing->Time = Connection->BlockedTimings.Scheduling.CumulativeTime;
+        if (Connection->BlockedTimings.Scheduling.LastStartTime != 0) {
+            Timing->Time +=
+                CxPlatTimeDiff64(Connection->BlockedTimings.Scheduling.LastStartTime, Now);
+        }
+        ++Timing;
+        Timing->Reason = QUIC_FLOW_BLOCKED_PACING;
+        Timing->Time = Connection->BlockedTimings.Pacing.CumulativeTime;
+        if (Connection->BlockedTimings.Pacing.LastStartTime != 0) {
+            Timing->Time +=
+                CxPlatTimeDiff64(Connection->BlockedTimings.Pacing.LastStartTime, Now);
+        }
+        ++Timing;
+        Timing->Reason = QUIC_FLOW_BLOCKED_AMPLIFICATION_PROT;
+        Timing->Time = Connection->BlockedTimings.AmplificationPort.CumulativeTime;
+        if (Connection->BlockedTimings.AmplificationPort.LastStartTime != 0) {
+            Timing->Time +=
+                CxPlatTimeDiff64(Connection->BlockedTimings.AmplificationPort.LastStartTime, Now);
+        }
+        ++Timing;
+        Timing->Reason = QUIC_FLOW_BLOCKED_CONGESTION_CONTROL;
+        Timing->Time = Connection->BlockedTimings.CongestionControl.CumulativeTime;
+        if (Connection->BlockedTimings.CongestionControl.LastStartTime != 0) {
+            Timing->Time +=
+                CxPlatTimeDiff64(Connection->BlockedTimings.CongestionControl.LastStartTime, Now);
+        }
+        ++Timing;
+        Timing->Reason = QUIC_FLOW_BLOCKED_CONN_FLOW_CONTROL;
+        Timing->Time = Connection->BlockedTimings.FlowControl.CumulativeTime;
+        if (Connection->BlockedTimings.FlowControl.LastStartTime != 0) {
+            Timing->Time +=
+                CxPlatTimeDiff64(Connection->BlockedTimings.FlowControl.LastStartTime, Now);
+        }
```

Comment:
We need to improve this. Right now this will return total timings for the lifetime of the entire connection. I _think_ the solution here would be to cache total cumulative time on the connection, when the stream is started, and then subtract that time from here when returned to the app.

Reply:
That's how perf counters usually work though, isn't it? They are absolute values, and the thread that is querying and using them subtracts the values from two queries. One advantage of this idempotent design is that multiple consumers of the perf counters can operate independently. That is less of an issue for this API where only the app owning the QUIC sockets can access the counters, but

1) that could feasibly change, and
2) it seems best to stick to the Standard Model

This is all assuming we haven't already added non-absolute perf counters somewhere in QUIC!

Reply:
Oh wait, I think I misunderstood you. You're saying the connection-wide timings really ought to be based on the time during which the stream existed, right? If so then yeah, I agree that's not right.

Reply:
Yeah, it's a problem.

Reply:
Yes, we need the connection's timings for only the lifetime of the stream.

Reply:
Done